### PR TITLE
Interactive Plotly Plotting for MMM - phase 2

### PR DESCRIPTION
**Pull request by isofer**
# PR Description: Interactive Plotly Plotting for MMM

## Description

This PR adds Plotly-based plotting for Marketing Mix Models. It uses `MMMSummaryFactory` to generate interactive plots from model results.

### What this adds

- `mmm.plot_interactive` property to access Plotly plots from the MMM class
- New plotting methods: `roas()`, `posterior_predictive()`, `saturation_curves()`, `adstock_curves()`
- HDI bands on line charts
- Extended auto-faceting to support up to 3 custom dimensions

### Technical approach

This PR extends the existing `MMMPlotlyFactory` class with new plotting methods and integrates it with the MMM class via a `plot_interactive` property.

Auto-faceting applies custom dimensions as: 1 dim → facet_col; 2 dims → facet_row + facet_col; 3 dims → adds line_dash.

### Design decisions

- Property-based access: `mmm.plot_interactive.method()` mirrors `mmm.plot.method()`
- Saturation curves and posterior predictive display values in original scale

### Core Implementation
- **`pymc_marketing/mmm/plot_interactive.py`** (~1000 lines added):
  - New methods added to `MMMPlotlyFactory`:
    - `roas()` - Return on Ad Spend bar charts
    - `posterior_predictive()` - Model fit line chart with HDI bands
    - `saturation_curves()` - Response curves
    - `adstock_curves()` - Decay curves
  - Extended auto-faceting logic
  - HDI band rendering for line charts

### MMM Integration
- **`pymc_marketing/mmm/multidimensional.py`**:
  - Added `plot_interactive` property to MMM class

### Bug Fixes
- **`pymc_marketing/mmm/summary.py`**:
  - Fixed posterior predictive to return values in original scale
  - Added `validate_data` parameter to `MMMSummaryFactory` to deal with changes to the data (while dealing with #2246)

### Tests
- **`tests/mmm/test_plot_interactive.py`** (~900 lines):
  - Tests for all plot types
  - Tests for auto-faceting with 0, 1, 2, and 3 custom dimensions
  - Edge case tests (missing data, invalid parameters)

## Demo

Demo notebook: https://nbviewer.org/gist/isofer/96139f10c9f73d288891cdbd85985d31

## How to Test

1. Load a fitted MMM model:
   ```python
   from pymc_marketing.mmm.multidimensional import MMM
   mmm = MMM.load("path/to/fitted_model.pm")
   ```

2. Test interactive plots:
   ```python
   fig = mmm.plot_interactive.posterior_predictive()
   fig.show()

   fig = mmm.plot_interactive.roas(frequency="yearly", color="channel", x="date")
   fig.show()

   fig = mmm.plot_interactive.saturation_curves(facet_col="brand")
   fig.show()

   fig = mmm.plot_interactive.adstock_curves()
   fig.show()
   ```

3. Run tests:
   ```bash
   pytest tests/mmm/test_plot_interactive.py -v
   ```

----
📚 Documentation preview 📚: https://pymc-marketing--2259.org.readthedocs.build/en/2259/

### CODE REVIEW

File: pymc_marketing/mmm/plot_interactive.py [hunk 1]
```diff
@@ -452,10 +752,793 @@ def contributions(
...
+        # Track if we've shown the legend (for non-color mode)
+        legend_shown = False
+
+        if not facet_dims:
+            # Non-faceted case: single iteration with row=1, col=1
+            facet_combinations: list[dict] = [{}]
+        else:
+            # Faceted case: get unique combinations using Narwhals
+            facet_df = nw_df.select(*facet_dims).unique()
+            facet_combinations = facet_df.to_native().to_dict("records")
```
[the remark this answers is not in the document]
Comment (isofer) [reply pymc_marketing/mmm/plot_interactive.py]: fixed.

File: pymc_marketing/mmm/plot_interactive.py [hunk 2]
```diff
@@ -452,10 +752,792 @@ def contributions(
...
+        else:
+            xaxis_title = "Spend (scaled)"
+
+        # Auto-detect faceting from custom dimensions
+        plotly_kwargs = self._apply_auto_faceting(
+            plotly_kwargs, auto_facet, single_dim_facet, supports_line_styling=True
+        )
+
+        return self._plot_curves(
+            df=nw_df.to_native(),
```
[the remark this answers is not in the document]
Comment (isofer) [reply pymc_marketing/mmm/plot_interactive.py:1408]: fixed

File: pymc_marketing/mmm/plot_interactive.py [hunk 3]
```diff
@@ -452,10 +752,793 @@ def contributions(
...
+
+        return self._plot_bar(
+            nw_df=nw_df,
+            y="mean",
+            hdi_prob=hdi_prob,
+            yaxis_title="ROAS",
+            frequency=frequency,
+            round_digits=round_digits,
+            **plotly_kwargs,
+        )
```
[the remark this answers is not in the document]
Comment (isofer) [reply pymc_marketing/mmm/plot_interactive.py:903]: fixed

File: pymc_marketing/mmm/plot_interactive.py [hunk 4]
```diff
@@ -229,50 +417,148 @@ def _apply_auto_faceting(self, plotly_kwargs: dict) -> dict:
...
+        ...     {}, auto_facet=False, supports_line_styling=True
+        ... )
+        >>> # Returns: {"line_dash": "country"} (uses line dash for differentiation)
         """
         plotly_kwargs = plotly_kwargs.copy()
 
-        # Skip if auto_facet disabled
-        if not self.auto_facet:
+        # Get custom dimensions
+        custom_dims = self.summary.data.custom_dims
```
[the remark this answers is not in the document]
Comment (isofer) [reply pymc_marketing/mmm/plot_interactive.py]: fixed

File: pymc_marketing/mmm/plot_interactive.py [hunk 5]
```diff
@@ -432,30 +804,789 @@ def contributions(
...
+                    x=facet_data.get_column(x).to_list(),
+                    lower=facet_data.get_column(lower_col).to_list(),
+                    upper=facet_data.get_column(upper_col).to_list(),
+                    name=f"{int(hdi_prob * 100)}% HDI" if hdi_prob else "HDI",
+                    opacity=hdi_opacity,
+                    showlegend=not legend_shown,  # Only show once in legend
+                    row=row_idx,
+                    col=col_idx,
+                )
+                legend_shown = True
```
[the remark this answers is not in the document]
Comment (isofer) [reply pymc_marketing/mmm/plot_interactive.py:1290]: fixed

File: pymc_marketing/mmm/plot_interactive.py [hunk 6]
```diff
@@ -432,30 +804,820 @@ def contributions(
...
+        nw_df = nw_df.sort("date")
+
+        # Create long-format DataFrame with both predicted and observed
+        id_cols = ["date", *self.custom_dims]
+        plot_df = nw_df.select(*id_cols, "mean", "observed").unpivot(
+            on=["mean", "observed"],
+            index=id_cols,
+            variable_name="series",
+            value_name="value",
+        )
```
[the remark this answers is not in the document]
Comment (isofer) [reply pymc_marketing/mmm/plot_interactive.py:993]: This is incorrect.

File: pymc_marketing/mmm/plot_interactive.py [hunk 7]
```diff
@@ -34,16 +108,49 @@
...
 from pymc_marketing.mmm.summary import MMMSummaryFactory
 
 # Type aliases matching MMMSummaryFactory for consistency
 ComponentType = Literal["channel", "control", "seasonality", "baseline"]
 
+# Default Plotly color sequence
+PLOTLY_COLORS = px.colors.qualitative.Plotly
+
+
+def _hex_to_rgba(color: str, opacity: float) -> str:
```
Comment (juanitorduz) [inline pymc_marketing/mmm/plot_interactive.py:134]: Can't we use some existing function like https://matplotlib.org/stable/api/_as_gen/matplotlib.colors.to_rgba.html (just curious)

File: pymc_marketing/mmm/plot_interactive.py [hunk 8]
```diff
@@ -55,37 +162,90 @@ class MMMPlotlyFactory:
...
-    >>> # Disable auto-faceting
-    >>> factory = MMMPlotlyFactory(mmm.summary, auto_facet=False)
-    >>> fig = factory.contributions(facet_col="country")
+    >>> # ROAS aggregated yearly, colored by date
+    >>> fig = mmm.plot_interactive.roas(frequency="yearly", color="date", x="channel")
+    >>> fig.show()
+
+    **Controlling faceting behavior**
+
+    For models with custom dimensions (geo, brand, etc.), auto_facet=True
```
Comment (juanitorduz) [inline pymc_marketing/mmm/plot_interactive.py:200]: Wasn't atufacet removed from the docs above?

File: pymc_marketing/mmm/plot_interactive.py [hunk 9]
```diff
@@ -432,30 +804,820 @@ def contributions(
...
+                            col=col_idx,
+                        )
+            else:
+                # Non-color mode: add single HDI band per facet
+                self._add_single_hdi_band(
+                    fig,
+                    x=facet_data.get_column(x).to_list(),
+                    lower=facet_data.get_column(lower_col).to_list(),
+                    upper=facet_data.get_column(upper_col).to_list(),
+                    name=f"{int(hdi_prob * 100)}% HDI" if hdi_prob else "HDI",
```
Comment (juanitorduz) [inline pymc_marketing/mmm/plot_interactive.py]: You can use f"{x:. 2%} of the time, it works every time." to display percentajes (2=# decimapl points) instead of hanling manually)

**Review by juanitorduz — APPROVED** [review]
Thanks @isofer LGTM. Just minor suggestions before we merge. I would be great to add a notebook (in a different PR) where we showcase all these functionalities by using the data and example https://www.pymc-marketing.io/en/stable/notebooks/mmm/mmm_multidimensional_example.html to compare the visualizations.

File: pymc_marketing/mmm/plot_interactive.py (hunk 8 above)
Comment (isofer) [reply pymc_marketing/mmm/plot_interactive.py:200]: it was removed from the doc above (which represent the default behavior), and added in the doc below with explanation on how to use it

File: pymc_marketing/mmm/plot_interactive.py (hunk 7 above)
Comment (isofer) [reply pymc_marketing/mmm/plot_interactive.py:134]: I didn't know about. good catch. thanks
fixed.

File: pymc_marketing/mmm/plot_interactive.py (hunk 9 above)
Comment (isofer) [reply pymc_marketing/mmm/plot_interactive.py]: nice. added.

File: pymc_marketing/mmm/plot_interactive.py [hunk 10]
```diff
@@ -17,33 +17,138 @@
...
+    opacity : float
+        Opacity value between 0 and 1
+
+    Returns
+    -------
+    str
+        RGBA color string (e.g., "rgba(99,110,250,0.3)")
+    """
+    rgba = np.array(mcolors.to_rgb(color)) * 256
+    return f"rgba({rgba[0]},{rgba[1]},{rgba[2]},{opacity})"
```
[the remark this answers is not in the document]
Comment (isofer) [reply pymc_marketing/mmm/plot_interactive.py:150]: fixed

**Review by juanitorduz — APPROVED** [review]
Great @isofer !
